python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -740,3 +742,83 @@ def with_reconstruction_metadata(self, state_metadata: Mapping[str, Any]) -> Sel
                 **state_metadata,
             },
         )
+
+
+def map_asset_specs(defs: Definitions, fn: Callable[[AssetSpec], AssetSpec]) -> "Definitions":
```

Comment:
Yeah that is definitely the biggest tradeoff / rough spot that's going on in this PR. Requiring the user to do the mapping of deps on their own would be the most "pure". However I went this direction because I believe it's prohibitively difficult for a user to implement this mapping on their own.

An alternative would be to also include a method like `map_asset_keys(self, fn: Callable[[AssetSpec], AssetKey])`.  `map_asset_keys` would handle deps, while `map_asset_specs` would not. Then if they want to customize both the asset keys and other attributes they'd need to do `defs.map_asset_keys(my_func_1).map_asset_specs(my_func_2)`. A couple potentially manageable downsides of this are:
- If there's some shared processing step that's used to determine the both asset key and other properties on the spec, annoying to need to repeat it in both functions
- Footgun of replacing the asset key within `map_asset_specs` (we could warn?)

> And what about AutomationConditions? Presumably any automation condition (or any object generally) that holds a reference to a key would have to be remapped.

Yeah if we introduce built-in AutomationConditions that target asset keys, it might make sense to remap them. Though users could define their own that we wouldn't know how to remap.

The other one to think about is jobs/sensors/schedules with asset selections that directly target asset keys.

I should have called out some of these complexities in the summary.

The best direction that I could think of is to take a pragmatic approach and draw the line the dependencies within the asset/check graph, be explicit about this, and expect users to handle other objects on their own. Very open to other ideas if there's something more general that you think would make more sense.  I do think that leaving it on users to handle all the remapping bookkeeping on their own would be a very painful experience.

Reply:
Is there a way that we can both (1) require the user to map the deps but (2) make this as low lift as possible?

I can think of a few ways to do this:

1. Provide a contextually available fn in the remapping fn

```python

class MapAssetSpecsContext:
    ...
    def remap_deps(self, spec: AssetSpec) -> AssetSpec:
        ...

def map_asset_specs(
    self,
    fn: Union[Callable[[AssetSpec], AssetSpec], Callable[[AssetSpec, MapAssetSpecsContext], AssetSpec]]
) -> "Definitions":
    ...

def my_custom_key_remapping(spec: AssetSpec, ctx: MapAssetSpecsContext) -> AssetSpec:
    custom_key = ...
    return ctx.remap_deps(spec._replace(key=custom_key))

power_bi_defs = power_bi_workspace.build_defs(...).map_asset_specs(my_custom_key_remapping)
```

This is attractive to me if we think we will end up wanting a context for other reasons down the line.

2. Provide a fn on `AssetSpec` which does this, either by marking the spec as to-be-remapped for the framework after it's returned, or remapping inline. In the latter case, the framework would probably end up calling the lambda twice per asset - once for each asset to assemble a key mapping and a second time to actually map the spec.

```python

def my_custom_key_remapping(spec: AssetSpec) -> AssetSpec:
    custom_key = ...
    return spec._replace(key=custom_key).remap_deps()

power_bi_defs = power_bi_workspace.build_defs(...).map_asset_specs(my_custom_key_remapping)
```

Reply:
I definitely agree the forcing the user to do the dep remapping is too much.

I like `defs.map_asset_keys(my_func_1).map_asset_specs(my_func_2)` and then we can hard error in `map_asset_specs` if someone tries to change the primary asset key.

What I like about that is that it makes it very clear what domain of asset keys the definitions object is at any one time. It might make more sense to remap using the old asset key scheme or the new one depending on the context.